TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -2,418 +2,594 @@
+  
+	std::vector<GAME_OBJECT_ID> CivvyExcludedTargets =
+	{
+		ID_CIVVY,
+		ID_VON_CROY,
+		ID_GUIDE,
+		ID_MONK1,
+		ID_MONK2,
+		ID_TROOPS
+	};
 
-	// TODO
 	enum CivvyState
 	{
 		// No state 0.
 		CIVVY_STATE_IDLE = 1,
-		CIVVY_STATE_WALK_FORWARD = 2,
-		CIVVY_PUNCH2 = 3,
-		CIVVY_AIM2 = 4,
-		CIVVY_WAIT = 5,
-		CIVVY_AIM1 = 6,
-		CIVVY_AIM0 = 7,
-		CIVVY_PUNCH1 = 8,
-		CIVVY_PUNCH0 = 9,
-		CIVVY_STATE_RUN_FORWARD = 10,
-		CIVVY_DEATH = 11,
-		CIVVY_CLIMB3 = 12,
-		CIVVY_CLIMB1 = 13,
-		CIVVY_CLIMB2 = 14,
-		CIVVY_FALL3 = 15
+		CIVVY_STATE_WALK = 2,
+		CIVVY_STATE_ATTACK_WALKING_PUNCH = 3, //Punch while walking
+		CIVVY_STATE_AIM_WALKING_PUNCH = 4,
+		CIVVY_STATE_WAIT = 5,
+		CIVVY_STATE_AIM_FAR_PUNCH = 6, //Punch from Idle to walk
+		CIVVY_STATE_AIM_CLOSE_PUNCH = 7, //Punch while Idle
+		CIVVY_STATE_ATTACK_FAR_PUNCH = 8,
+		CIVVY_STATE_ATTACK_CLOSE_PUNCH = 9,
+		CIVVY_STATE_RUN = 10,
+		CIVVY_STATE_DEATH = 11,
+		CIVVY_STATE_CLIMB_2CLICKS = 12,
+		CIVVY_STATE_CLIMB_3CLICKS = 13,
+		CIVVY_STATE_CLIMB_4CLICKS = 14,
+		CIVVY_STATE_FALL_4CLICKS = 15
 	};
 
-	// TODO
 	enum CivvyAnim
 	{
+		CIVVY_ANIM_WALK = 0,
+		CIVVY_ANIM_WALKING_PUNCH_PREPARE = 1,
+		CIVVY_ANIM_WALKING_PUNCH_TO_IDLE = 2,
+		CIVVY_ANIM_WALKING_PUNCH_TO_WALK = 3,
+		CIVVY_ANIM_WALKING_PUNCH_ATTACK = 4,
+		CIVVY_ANIM_IDLE_TO_WALK = 5,
 		CIVVY_ANIM_IDLE = 6,
-
+		CIVVY_ANIM_WALKING_PUNCH_CANCEL = 7,
+		CIVVY_ANIM_WALK_TO_IDLE_LEFT = 8,
+		CIVVY_ANIM_CLOSE_PUNCH_CANCEL = 9,
+		CIVVY_ANIM_CLOSE_PUNCH_PREPARE = 10,
+		CIVVY_ANIM_FAR_PUNCH_CANCEL = 11,
+		CIVVY_ANIM_FAR_PUNCH_PREPARE = 12,
+		CIVVY_ANIM_CLOSE_PUNCH_TO_IDLE = 13,
+		CIVVY_ANIM_FAR_PUNCH_TO_IDLE = 14,
+		CIVVY_ANIM_FAR_PUNCH_TO_WALK = 15,
+		CIVVY_ANIM_CLOSE_PUNCH_ATTACK = 16,
+		CIVVY_ANIM_FAR_PUNCH_ATTACK = 17,
+		CIVVY_ANIM_IDLE_TO_WAIT = 18,
+		CIVVY_ANIM_WAIT_TO_IDLE = 19,
+		CIVVY_ANIM_WAIT = 20,
+		CIVVY_ANIM_RUN = 21,
+		CIVVY_ANIM_WALK_TO_RUN = 22,
+		CIVVY_ANIM_RUN_TO_WALK_LEFT = 23,
+		CIVVY_ANIM_RUN_TO_IDLE = 24,
+		CIVVY_ANIM_IDLE_TO_RUN = 25,
 		CIVVY_ANIM_DEATH = 26,
-		CIVVY_CLIMB3_ANIM = 27,
-		CIVVY_CLIMB1_ANIM = 28,
-		CIVVY_CLIMB2_ANIM = 29,
-		CIVVY_FALL4_ANIM = 30
+		CIVVY_ANIM_CLIMB_4CLICKS = 27,
+		CIVVY_ANIM_CLIMB_2CLICKS = 28,
+		CIVVY_ANIM_CLIMB_3CLICKS = 29,
+		CIVVY_ANIM_FALL_4CLICKS = 30,
+		CIVVY_ANIM_RUN_TO_WALK_RIGHT = 31,
 	};
 
 	void InitialiseCivvy(short itemNumber)
 	{
-		auto* item = &g_Level.Items[itemNumber];
+		auto& item = g_Level.Items[itemNumber];
 
 		InitialiseCreature(itemNumber);
-		SetAnimation(item, CIVVY_ANIM_IDLE);
+		SetAnimation(&item, CIVVY_ANIM_IDLE);
+	}
+
+
+	//In the future, if this function works, it could be taken to a more neutral place to be reused by other ally entities.
+	ItemInfo* CivvyFindNearTarget(short itemNumber, std::vector<GAME_OBJECT_ID>& excludedTargetList, int rangeDetection)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		ItemInfo* result = nullptr;
+
+		auto closetsDistance = FLT_MAX;
```

Comment:
Closest, not closets.

Reply:
Done (Although I let it with lower capital, **closestDistance** because is a local var).